File: analytics/scala/src/test/scala/com/koverse/example/spark/TestUtils.scala
package com.koverse.example.spark

import com.koverse.thrift.client.{Client, ClientConfiguration}
import com.koverse.thrift.{TConfigValue, TNotFoundException}
import com.koverse.thrift.collection.TCollection
import com.koverse.thrift.dataflow.{TImportFlow, TImportFlowType, TJobAbstract, TSource, TTransform, TTransformInputDataWindowType, TTransformScheduleType}
import java.io.FileInputStream
import java.util
import java.util.{HashMap, HashSet, Map, Properties, Set}

import org.apache.thrift.TException

import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.collection.mutable.ListBuffer

trait TestUtils {

  val client = connect

  protected def connect: Client = {
    val rootPath = Thread.currentThread.getContextClassLoader.getResource("").getPath
    val appConfigPath = rootPath + "client.properties"
    val appProperties = new Properties
    appProperties.load(new FileInputStream(appConfigPath))
    val host = appProperties.getProperty("koverse.host")
    val name = appProperties.getProperty("client.name")
    val secret = appProperties.getProperty("client.secret")
    if (host.isEmpty || name.isEmpty || secret.isEmpty)
      throw new IllegalArgumentException("You must update the client.properties file before running this example.")
    val config = ClientConfiguration.builder.host(host).clientName(name).clientSecret(secret).build
    new Client(config)
  }

  protected def createDataSet(name: String): TCollection = client.createDataSet(name)

  protected def setUpImport(importDataSet: TCollection, pages: String): TCollection = {
    val wikipediaSource = new TSource
    wikipediaSource.setName(importDataSet.getName + " Wikipedia WikipediaSourceUtil")
    wikipediaSource.setTypeId("wikipedia-pages-source")
    val importOptions = new util.HashMap[String, String]
    importOptions.put("pageTitleListParam", pages)
    wikipediaSource.setParameters(importOptions)
    // this will fill out the ID of the source
    val wikipediaSourceId = client.createSource(wikipediaSource).getSourceId
    var importFlow = new TImportFlow
    importFlow.setSourceId(wikipediaSourceId)
    importFlow.setDataCollectionId(importDataSet.getId)
    importFlow.setType(TImportFlowType.MANUAL)
    importFlow = client.createImportFlow(importFlow)
    // save import flow id back to dataset
    val importFlowId:java.lang.Long = importFlow.getImportFlowId
    val importFlowIds = List(importFlowId)
    importDataSet.setImportFlowIds(importFlowIds.asJava)
    importDataSet.setImportFlowId(importFlow.getImportFlowId)
    client.updateDataSet(importDataSet)
  }

  protected def configureAndSaveTransform(transformType: String, transformOptions: util.Map[String, TConfigValue]) = {
    println("setting up transform: " + transformType)
    val transform = new TTransform
    transform.setType(transformType)
    transform.setDisabled(false)
    // setting the schedule type to AUTOMATIC means the transform will run
    // whenever there is new data to process
    transform.setScheduleType(TTransformScheduleType.AUTOMATIC)
    transform.setInputDataWindowType(TTransformInputDataWindowType.NEW_DATA)
    transform.setReplaceOutputData(false)
    transform.setInputDataSlidingWindowOffsetSeconds(0)
    transform.setInputDataSlidingWindowSizeSeconds(0)
    // configure the transform to process text in the "article" field
    transform.setParameters(transformOptions)
    client.createTransform(transform)
  }

  protected def executeAndMonitorImportFlow(dataSetId: String, transformDataSetId: String) = {
    val dataSet = client.getDataSet(dataSetId)
    executeImportFlow(client, dataSet)
    // wait for import to complete
    checkJobProgress(dataSet.getId, transformDataSetId)
  }

  private def executeImportFlow(client: Client, dataSet: TCollection) = {
    val importFlowIds = dataSet.getImportFlowIds.asScala
    // start the import
    importFlowIds.foreach( id => {
      println(s"executing data flow for ${dataSet.getName} $id")
      client.executeImportFlow(id)
    })
  }

  private def checkJobProgress(dataSetId: String, transformDataSetId: String) = {
    val jobIds = new util.HashSet[Long]
    val jobStatus = new util.HashMap[String, String]
    println("Monitoring jobs for " + dataSetId)
    // we'll wait until the import job we requested starts
    // but only wait 1 minute in case it's finished before we get here
    try {
      var jobs = client.getAllActiveJobs(dataSetId)
      jobs.addAll(client.getAllActiveJobs(transformDataSetId))
      while ( {
        jobs.isEmpty
      }) {
        Thread.sleep(2000)
        System.out.println("waiting for jobs to start for " + dataSetId)
        jobs = client.getAllJobs(dataSetId)
        jobs.addAll(client.getAllActiveJobs(transformDataSetId))
      }
    } catch {
      case ex@(_: InterruptedException | _: TException) =>
        System.out.println(ex.getMessage)
    }
    val allJobs = client.getAllJobs(dataSetId).asScala
    for (alljob <- allJobs) {
      if ((alljob.getStatus eq "success") || (alljob.getStatus eq "error")) jobIds.add(alljob.getId)
      else {
        var allActiveJobs = client.getAllActiveJobs(dataSetId)
        allActiveJobs.addAll(client.getAllActiveJobs(transformDataSetId))
        // while there are jobs running we want to keep updating our map of statuses
        while ( {
          !allActiveJobs.isEmpty
        }) {
          Thread.sleep(5000)
          allActiveJobs = client.getAllActiveJobs(dataSetId)
          allActiveJobs.addAll(client.getAllActiveJobs(transformDataSetId))
          import scala.collection.JavaConversions._
          for (job <- allActiveJobs) {
            jobIds.add(job.getId)
            // keep for troubleshooting
            //            System.out.println(String.format(
            //            "%s %d %s:%s %f", dataSetId, job.getId(), job.getType(), job.getStatus(), job.getProgress()));
          }
          for (jobId <- jobIds) {
            val job = client.getJob(jobId)
            if (job.getStatus == "error")
              println(s"${job.getType} ${job.getId} completed with status error: ${job.getErrorDetail}")
          }
        }
      }
    }
    println("Jobs completed.")

    val it = jobIds.iterator()
    while(it.hasNext) {
      val id = it.next()
      try {
        val tJobAbstract = client.getJob(id)
        println("Job " + id + ", " + tJobAbstract.getStatus)
        jobStatus.put(tJobAbstract.getType.toString, tJobAbstract.getStatus)
      } catch {
        case ex: TException =>
          jobStatus.put("ERROR", "exception")
      }
    }
  }


  protected def tearDownDataFlow(dataFlowName: String) = {
    try {
      val dataSet = client.getDataSetByName(dataFlowName)
      // deleting this data set will delete associated sources and transforms
      println("Deleting wikipedia pages data set ...")
      client.deleteDataSet(dataSet.getId)
    } catch {
      case tnfe: TNotFoundException =>
      // nothing to remove .. so skip
    }

    try {
      val keywordDataSet = client.getDataSetByName(dataFlowName + " Keywords")
      println("Deleting keyword data set ...")
      client.deleteDataSet(keywordDataSet.getId)
    } catch {
      case tnfe: TNotFoundException =>
      // nothing to remove
    }
    println(String.format("Done tearing down data flow: %s", dataFlowName))
  }

}

File: analytics/java/src/test/java/com/koverse/examples/analytics/TestUtils.java
package com.koverse.examples.analytics;

import static com.koverse.com.google.common.collect.Lists.newArrayList;

import com.koverse.thrift.TConfigValue;
import com.koverse.thrift.TNotFoundException;
import com.koverse.thrift.client.Client;
import com.koverse.thrift.client.ClientConfiguration;
import com.koverse.thrift.collection.TCollection;
import com.koverse.thrift.dataflow.TImportFlow;
import com.koverse.thrift.dataflow.TImportFlowType;
import com.koverse.thrift.dataflow.TJobAbstract;
import com.koverse.thrift.dataflow.TSource;
import com.koverse.thrift.dataflow.TTransform;
import com.koverse.thrift.dataflow.TTransformInputDataWindowType;
import com.koverse.thrift.dataflow.TTransformScheduleType;

import org.apache.thrift.TException;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;

public class TestUtils {
  Client client;

  TestUtils() {
    try {
      client = connect();
    } catch (IOException | TException ex) {
      System.out.println(ex.getMessage());
    }
  }

  protected Client connect() throws IOException, TException {

    String rootPath = Thread.currentThread().getContextClassLoader().getResource("").getPath();
    String appConfigPath = rootPath + "client.properties";

    Properties appProperties = new Properties();
    appProperties.load(new FileInputStream(appConfigPath));

    String host = appProperties.getProperty("koverse.host");
    String name = appProperties.getProperty("client.name");
    String secret = appProperties.getProperty("client.secret");

    if(host.isEmpty() || name.isEmpty() || secret.isEmpty()) {
      throw new IllegalArgumentException("You must update the client.properties file before running this example.");
    }

    ClientConfiguration config =
        ClientConfiguration.builder().host(host).clientName(name).clientSecret(secret).build();

    return new Client(config);
  }

  protected TCollection createDataSet(String name) throws TException{
    return client.createDataSet(name);
  }

  protected TCollection setUpImport(TCollection importDataSet, String pages) throws TException {
    TSource wikipediaSource = new TSource();
    wikipediaSource.setName(importDataSet.getName() + " Wikipedia WikipediaSourceUtil");
    wikipediaSource.setTypeId("wikipedia-pages-source");

    Map<String, String> importOptions = new HashMap<>();
    importOptions.put("pageTitleListParam", pages);
    wikipediaSource.setParameters(importOptions);

    // this will fill out the ID of the source
    wikipediaSource = client.createSource(wikipediaSource);

    TImportFlow importFlow = new TImportFlow();

    importFlow.setSourceId(wikipediaSource.getSourceId());
    importFlow.setDataCollectionId(importDataSet.getId());
    importFlow.setType(TImportFlowType.MANUAL);
    importFlow = client.createImportFlow(importFlow);

    // save import flow id back to dataset
    List<Long> importFlowIds = newArrayList(importFlow.getImportFlowId());
    importDataSet.setImportFlowIds(importFlowIds);
    importDataSet.setImportFlowId(importFlow.getImportFlowId());

    return client.updateDataSet(importDataSet);
  }

  protected void configureAndSaveTransform(
      String transformType,
      Map<String, TConfigValue> transformOptions)
      throws TException {

    System.out.println("setting up transform: " + transformType);
    TTransform transform = new TTransform();

    transform.setType(transformType);
    transform.setDisabled(false);
    // setting the schedule type to AUTOMATIC means the transform will run
    // whenever there is new data to process
    transform.setScheduleType(TTransformScheduleType.AUTOMATIC);
    transform.setInputDataWindowType(TTransformInputDataWindowType.NEW_DATA);
    transform.setReplaceOutputData(false);
    transform.setInputDataSlidingWindowOffsetSeconds(0);
    transform.setInputDataSlidingWindowSizeSeconds(0);

    // configure the transform to process text in the "article" field

    transform.setParameters(transformOptions);

    client.createTransform(transform);

  }

  protected void executeAndMonitorImportFlow(String dataSetId, String transformDataSetId)
      throws TException, InterruptedException {

    TCollection dataSet = client.getDataSet(dataSetId);
    executeImportFlow(client, dataSet);
    // wait for import to complete
    checkJobProgress(dataSet.getId(), transformDataSetId);

  }

  private void executeImportFlow(Client client, TCollection dataSet) throws TException {
    List<Long> importFlowIds = dataSet.getImportFlowIds();

    // start the import


    importFlowIds.forEach (
      id -> {
        try {
          System.out.println(String.format("executing data flow for %s %d", dataSet.getName(), id));
          client.executeImportFlow(id);
        } catch (TException e) {
          System.out.println(e.getMessage());
        }
      });

  }

  private Map<String, String> checkJobProgress(String dataSetId, String transformDataSetId) throws TException, InterruptedException {
    Set<Long> jobIds = new HashSet<>();
    Map<String, String> jobStatus = new HashMap<>();

    System.out.println("Monitoring jobs for " + dataSetId);
    // we'll wait until the import job we requested starts
    // but only wait 1 minute in case it's finished before we get here

    try {
      List<TJobAbstract> jobs = client.getAllActiveJobs(dataSetId);
      jobs.addAll(client.getAllActiveJobs(transformDataSetId));
      while (jobs.isEmpty()) {
        Thread.sleep(2000);
        System.out.println("waiting for jobs to start for " + dataSetId);
        jobs = client.getAllJobs(dataSetId);
        jobs.addAll(client.getAllActiveJobs(transformDataSetId));
      }
    } catch (InterruptedException | TException ex) {
      System.out.println(ex.getMessage());
    }

    List<TJobAbstract> allJobs = client.getAllJobs(dataSetId);
    for (TJobAbstract alljob: allJobs) {
      if (alljob.getStatus() == "success" || alljob.getStatus() == "error") {
        jobIds.add(alljob.getId());
      } else {

        List<TJobAbstract> allActiveJobs = client.getAllActiveJobs(dataSetId);
        allActiveJobs.addAll(client.getAllActiveJobs(transformDataSetId));
        // while there are jobs running we want to keep updating our map of statuses
        while (!allActiveJobs.isEmpty()) {
          Thread.sleep(5000);
          allActiveJobs = client.getAllActiveJobs(dataSetId);
          allActiveJobs.addAll(client.getAllActiveJobs(transformDataSetId));
          for (TJobAbstract job : allActiveJobs) {
            jobIds.add(job.getId());
            // keep for troubleshooting
            //            System.out.println(String.format(
            //            "%s %d %s:%s %f", dataSetId, job.getId(), job.getType(), job.getStatus(), job.getProgress()));
          }
          for (Long jobId : jobIds) {
            TJobAbstract job = client.getJob(jobId);
            if (job.getStatus().equals("error")) {
              System.out.println(
                  String.format("%s %s completed with status error: %s", job.getType(), job.getId(), job.getErrorDetail()));
            }
          }
        }
      }
    }

    System.out.println("Jobs completed.");

    jobIds.forEach((id) -> {
      String finalStatus;
      String jobType;
      try {
        TJobAbstract tJobAbstract = client.getJob(id);
        finalStatus = tJobAbstract.getStatus();
        jobType = tJobAbstract.getType().name();
        System.out.println("Job " + id + ", " + finalStatus);
      } catch (TException ex) {
        finalStatus = "exception";
        jobType = "ERROR";
      }
      jobStatus.put(jobType, finalStatus);
    });

    return jobStatus;
  }


  protected void tearDownDataFlow(String dataFlowName) throws TException {

    try {
      TCollection dataSet = client.getDataSetByName(dataFlowName);

      // deleting this data set will delete associated sources and transforms
      System.out.println("Deleting wikipedia pages data set ...");
      client.deleteDataSet(dataSet.getId());
    } catch (TNotFoundException tnfe) {
      // nothing to remove .. so skip
    }

    try {
      TCollection keywordDataSet = client.getDataSetByName(dataFlowName + " Keywords");

      System.out.println("Deleting keyword data set ...");
      client.deleteDataSet(keywordDataSet.getId());
    } catch (TNotFoundException tnfe) {
      // nothing to remove
    }

    System.out.println(String.format("Done tearing down data flow: %s", dataFlowName));
  }


}
